sw/device/silicon_creator/rom/rom.c

```diff
@@ -130,10 +133,27 @@ OT_WARN_UNUSED_RESULT
 static rom_error_t rom_init(void) {
   CFI_FUNC_COUNTER_INCREMENT(rom_counters, kCfiRomInit, 1);
   sec_mmio_init();
-  // Initialize pinmux configuration so we can use the UART.
-  pinmux_init();
-  // Configure UART0 as stdout.
-  uart_init(kUartNCOValue);
+
+  uint32_t reset_reasons = rstmgr_reason_get();
```

Comment:
Is there any meaningful difference if we do the following?

```c 
waking_from_sleep = kHardenedBoolFalse;
if (launder32(reset_reasons) & (1 << kRstmgrReasonLowPowerExit)) {
  waking_from_sleep = kHardenedBoolTrue;
}
```

I can't think of a case where the attacker could benefit from skipping the branch, but if you can think of one, then it may be worth it to add a redundant check later on.

```c
if (launder32(rstmgr_reason_get()) & (1 << kRstmgrReasonLowPowerExit)) {
  HARDENED_CHECK_EQ(waking_from_sleep, kHardenedBoolTrue);
}
```

Reply:
If we wake up from low power mode, but before executing to the point of claring the reset reason register, the external POR_N is asserted (or we hit a watchdog, undervoltage or any other reset source), then as execution reaches this point we might find TWO or more bits in the register set, one of them being low power exit.  We want to treat the above situations as hard reset (not waking up from low power), since even the always on IP blocks would have had their state reset.  So that is why my logic seem to look at every bit except the "low power exit" one.

Reply:
Hehe, just being pedantic here, but POR_N specifically would reset the reset reason register (rstmgr will be reset). There would only be the one bit set in that particular case.

Reply:
Good to know. But a watchdog reset or plain software reset (trap) while waking up from low power would result in AON IPs having their state reset, and a combination of bits in the reset cause register, so I still think that the logic of looking at every other bit in the register is valid.